README.md
# Cookie Monster [![Build Status](https://travis-ci.org/Anahkiasen/cookie-monster.png?branch=master)](https://travis-ci.org/Anahkiasen/cookie-monster)

Cookie Monster is a plugin you can load into Cookie Clicker, that offers a wide range of tools and statistics to enhance the game.
It is **not** a cheat interface – although it does offer helpers for golden cookies and such, everything can be toggled off at will to only leave how much information you want.

This is a helper, and it is here to help you at _whichever_ degree you want, if you only need some help shortening long numbers, it does that. If you need to be accompanied
by hand to pick the best buildings to buy, it does that, but **everything is an option**.

### Current version

You can see the current version, and a full history of all versions and what they changed by consulting the [Releases page](https://github.com/Anahkiasen/cookie-monster/releases).

### What it does

At its core, Cookie Monster computes two indexes on both buildings and upgrades :

* **Base Cost per Income (BCI)** : indicates how much a building is worth by comparing how much it costs to how much it will earn back
* **Return on investment (ROI)** : indicates how "interesting" a building is, for example, if it reduces the time to buy a more advanced building

Cookie Monster also indicates the time left before being able to buy an upgrade or building, and takes it into consideration. It will take *everything* in consideration, meaning if buying a building also unlocks an achievement which boosts your income, which unlocks an achievement, it will know and highlight that building's value.

Both these indexes are computed for buildings and upgrades. If the relevant option is enabled, it will color-code each of them based on their value :

* Light Blue : (upgrades) This item has a better BCI than any building
* Green      : This item has the best BCI
* Yellow     : This item is not the best, but it is closer to best than it is to worst
* Orange     : This item is not the worst, but it is closer to worst than it is to best
* Red        : This item has the worst BCI
* Purple     : (upgrades) This item has a worst BCI than any building
* Grey       : (upgrades) This item has not been calculated and/or cannot be calculated due to no definitive worth.

Note : for both these indexes, **lower is better**, meaning a building with a BCI of 1 is more interesting that one with a BCI of 3.

### What it doesn't do

Most likely you'll find items in grey like Golden Cookie upgrades, clicking upgrades – everything that doesn't earn you a direct bonus to your income will display as grey.
This means the following upgrades are **not** taken into account by Cookie Monster :

* Plastic mouse
* Iron mouse
* Titanium mouse
* Adamantium mouse
* Unobtainium mouse
* Lucky day
* Serendipity
* Get lucky
* Elder Pledge
* Sacrificial rolling pins
* **etc.**

Do note though that, although these upgrades have no direct value, if buying them earns you an achievement of some sort which in return gives you milk and income, Cookie Monster **will** display that value.

## Using

### Bookmarklet

Simply use like the previous plugin, create a bookmark and set this as the URL. Then when in the game, click on that bookmark.

```js
javascript: (function () {
  var jA = document.createElement('script');
  jA.setAttribute('type', 'text/javascript');
  jA.setAttribute('src', 'http://cookie-monster.autopergamene.eu/cookie-monster.min.js?' + new Date().getTime());
  document.body.appendChild(jA);
}());
```

If (for some reason) the above doesn't work, trying pasting everything after the `javascript:` bit into your browser's console.

### Userscript

If you'd rather use the plugin as a script via, for example, _Greasemonkey_ or _Tampermonkey_, you can use the following script, which will automatically load _Cookie Monster_ every time the original game loads.
You may need to specify `http://orteil.dashnet.org/cookieclicker/` when asked for a _namespace_ or _Includes_. For how to add an userscript to your browser, refer to your browser/plugin's documentation as the method changes for each one.

```js
// ==UserScript==
// @name        Cookie Monster
// @namespace   Cookie
// @include     http://orteil.dashnet.org/cookieclicker/
// @version     1
// @grant       none
// ==/UserScript==

var init  = Game.Init;
Game.Init = function() {
  init();
  (function () {
    var jA = document.createElement('script');
    jA.setAttribute('type', 'text/javascript');
    jA.setAttribute('src', 'http://cookie-monster.autopergamene.eu/cookie-monster.min.js?' + new Date().getTime());

    document.body.appendChild(jA);
  }());
}
```

## Bugs and suggestions

Any bug or suggestion should be **opened as an issue** [in the repository](https://github.com/Anahkiasen/cookie-monster/issues) for easier tracking. This allows me to close issues once they're fixed.

Before submitting a bug, make sure to give a shot at the latest version of the plugin on the `develop` branch. For this, simply replace `http://cookie-monster` by `http://dev.cookie-monster` in the bookmarklet above.
If the bug is still here, you can submit an issue for it.

All suggestions are welcome, even the smallest ones.

## Contributing

All edits must be made to the [develop branch](https://github.com/Anahkiasen/cookie-monster/tree/develop), that's where the latest version of the code is.

To contribute to the project, you'll find the files you want to edit in `src/`, split by what the methods are related to. If you're not familiar with [Git](http://git-scm.com/), the simplest way to contribute is to go into a file, click the _Edit_ button at the top, edit the code, and click on **"Propose file change"**.

### Setup the repository

If you are familiar with git, simply click the **"Fork"** button at the top of the page. You'll be redirected to your fork of the repository. Copy its address in the right side bar. Then clone the repository and install the dependencies :

    # Clone repository
    git clone https://github.com/YOURNAME/cookie-monster.git
    cd cookie-monster

    # Install dependencies
    npm install
    bundle install

You'll need NPM for that, if you don't have it, you can find a one-click installer on [NodeJS's website](http://nodejs.org/) or if you're more familiar with development and have per example Homebrew, you can just do `brew install node` or on a Linux distrobution such as Ubuntu, `apt-get install npm nodejs-legacy` or equivalent. Other dependencies (as named on Ubuntu 13.10) include `ruby2.0` and `ruby2.0-dev`.

If you don't have `bundle` run `gem install bundler`.

### Editing

Make the edits you want in the `src/` folder.

### Compiling and sending

Then rebuild the assets by running Grunt :

    ./node_modules/.bin/bower install
    ./node_modules/.bin/grunt

Test your changes by hosting a local http server with the contents of the dist directory and using a modified bookmarklet such as

```js
javascript: (function () {
  var jA = document.createElement('script');
  jA.setAttribute('type', 'text/javascript');
  jA.setAttribute('src', 'http://localhost:8000/cookie-monster.min.js?' + new Date().getTime());
  document.body.appendChild(jA);
}());
```

Next you can commit your changes and push :

    git commit -m "Edited some stuff" src
    git push origin master

Once that is done, go back on Github where your fork is, and create a pull request from there.

## Contributors

* **[Raving_Kumquat](http://cookieclicker.wikia.com/wiki/User:Raving_Kumquat)** : Original author
* **[Maxime Fabre](https://github.com/Anahkiasen)** : Current maintainer
* **Alderi Tokori** : ROI calculations


CHANGELOG.md
## 1.045 Versions

### 1.045.01 _(2014-04-24)_

* Bugfixes, bugfixes everywhere

## 1.041 Versions

### 1.041.03 _(2014-02-21)_

* Bugfixes

### 1.041.02 _(2014-02-15)_

* Bugfixes

### 1.041.01 _(2014-02-14)_

* Added scientific notations for numbers
* Added "Bonus to Clicking income" stat on clicking upgrades' tooltips
* Added "Percentage of current income" in tooltips
* Fixed Santa Upgrades position on the screen
* Prevent tooltips from going out of the screen

## 1.040 Versions

### 1.040.11 _(2014-01-05)_

* Tweak formula for computing the "grey bars"'s width
* Add time left next to unattained lucky tresholds in Statistics
* Add a new "Estimates" section in Statistics, time estimates are computed for can be set in options
* Slight fix to computed pledge costs
* Fixed "Lucky Reward (CUR)" not updating during frenzies

### 1.040.10 _(2014-01-02)_

* Added timer next to the Lucky Alert deficits to show time before the deficit is filled
* Fixed Reindeer value in statistics
* Added statistics about whether it's better to buy pledge or make a covenant
* Added a _grey_ color for upgrades, to differentiate items with low BCI/no BCI
* Added support for multiple title modifiers (and Reindeer title modifier added with it)

### 1.040.09 _(2014-01-02)_

* Finally fix BCI going crazy during frenzies
* Moved all assets to private server (you need to update your bookmarklet)
* Upgrades that impact negatively the income (eg. Elder Covenant) now appear in purple
* Fixed some bars-related bugs
* Wrinklers statistics are now hidden if no Wrinklers active

### 1.040.08 _(2014-01-01)_

* Take into account how an object/upgrade will affect the income to compute the Lucky Cookies deficit alerts
* Fix frenzies not affecting BCI
* Take into account some more upgrades like Elder Covenant
* Also fixed the BCI of some upgrades that displayed NaN or an inferior number

### 1.040.07 _(2014-01-01)_

* Fixed a bug where some tooltips weren't being displayed on newly unlocked upgrades
* Added **ROI** (Return on Investment) option as an alternative to **BCI** (Base Cose per Income) for buildings
* Loading a save or resetting the game now properly refreshes Cookie Monster
* Fixed BCI going crazy during frenzies
* Small tweak to the wrinklers revenues formula

### 1.040.06 _(2013-12-31)_

* Added Wrinklers statistics
* Changed Reindeer notification sound
* Fixed timer bars (reindeers, etc.) width being miscalculated
* Fixed a bug where having twice the same frenzy would mess with the bars

### 1.040.05 _(2013-12-31)_

* Add some upgrades that were still not taken into account
* Added back the "grey timeout" for Next Cookie/Reindeed bars
* Huge performances work, CM should run a lot smoother now (ie. more than the original)
* Fix tooltips displaying full time instead of compressed time

### 1.040.04 _(2013-12-30)_

* Take into account new upgrades
* Fixed a bug where the bottom bar would appear even if disabled
* Fixed a bug where some buff bars would appears as uncolored

### 1.040.03 _(2013-12-30)_

* Add Reindeer timer bars and sound (linked to Golden Cookie options for now)

### 1.040.02 _(2013-12-29)_

* Added back colorblind mode as an option
* Fixed a miscalculation in Bonus Income

### 1.040.01 _(2013-12-23)_

* Fixed some bugs related to the Grandmapocalypse
* Fixed a bug where the Golden Cookie sound would crash the application

## 1.038 Versions

### 1.038.01 _(2013-10-15)_

* Updated the buff timers to have a width relative to that of the Cookie Timer Bar
* Added the calculations for the new Upgrades for Heavenly Chips
* Added the calculation for the new Heavenly Chips Achievement "Wholesome"
* Added the missing calculations for the Grandma Achievements "Friend of the ancients" and "Ruler of the ancients"
* Fixed an issue that was causing 9 additional Buildings to be bought on the first purchase made after loading Cookie Monster _~ "don't even ask..."_
* Fixed "Session started" format to display as "days, hours, minutes, seconds" again
* Fixed and updated the "Next Cookie Timer"
	* Added a Grey part to the timer bar to depict the time left before a cookie has a chance of spawning
	* Note that, with the new update to the game, golden cookies now have a minimum and maximum time before they spawn (this makes cookie spawns unpredictable between the 2 values)

## 1.037 Versions

### 1.037.01 _(2013-10-02)_

* **Lucky Alert** will now tell you if you are currently under the required amount
* Updated the Kitten Overseers calculation that was changed in v.1.037
* Fixed "Game started" format to display as "days, hours, minutes, seconds" again
* Fixed an issue that was causing NaN values after the "Sextillion Fingers" upgrade

## 1.036 Versions

### 1.036.08 _(2013-09-24)_

* Added a new setting to the "**Cookie Monster Settings**" section on the Menu page
	* **Upgrade Display** - _"Changes how the store displays Upgrades"_
		* **Normal **- The default display
		* **All **- Expands the store to show all Upgrades
		* **None **- Collapses the store to hide all Upgrades
* Changed the options for the **Short Numbers** setting
	* **Short Numbers ON (A)** - Progression: _M, B, T, Qa, Qi, Sx, Sp, Oc, No, Dc_
	* **Short Numbers ON (B)** - Progression: _M, G, T, P, E, Z, Y, Oc, No, Dc_
	* **Short Numbers OFF** - No progression.
* Updated the display of "days, hours, minutes, seconds" to only show significant values
	* **i.e.** "0 days, 0 hours, 2 minutes, 0 seconds" will now be displayed as "2 minutes, 0 seconds"
* Fixed a hosting issue for Cookie Monster images and sounds _(you may have already noticed this change)_
* Fixed an issue that was causing Cookie Monster to think "Enhancer", "Augmenter", and "Upgrader" were never Achieved
* Fixed an issue that was causing Lucky Alerts for Upgrades to show a Deficit of "NaN"

### 1.036.07 _(2013-09-20)_

* Added a new setting to the "**Cookie Monster Settings**" section on the Menu page

src/tooltips.js
/**
 * Save the currently available tooltips
 *
 * @return {void}
 */
CookieMonster.saveTooltips = function() {
	Game.UpgradesById.forEach(function (upgrades, key) {
		CookieMonster.tooltips[key] = upgrades.desc;
	});
	Game.ObjectsById.forEach(function (building, key) {
		CookieMonster.buildingTooltips[key] = building.desc;
	});
};

CookieMonster.setupTooltips = function() {
	var needsRebuild = false;

	Game.UpgradesById.forEach(function (upgrade, key) {
		for (var upgradeKey = 0; upgradeKey < CookieMonster.upgradeCount; upgradeKey++) {
			if (CookieMonster.checkUpgrade(upgradeKey, key, true)) {
				upgrade.desc = CookieMonster.manageTooltips(upgradeKey, key, true, false);
				needsRebuild = true;
				break;
			}
		}
		if (upgrade.bought && upgrade.desc !== CookieMonster.tooltips[key]) {
			upgrade.desc = CookieMonster.tooltips[key];
			needsRebuild = true;
		}
	});

	if (needsRebuild) {
		Game.RebuildUpgrades();
	}
};

/**
 * Update one or more types of tooltips
 *
 * @param {string} which [upgrades,objects,all]
 *
 * @return {void}
 */
CookieMonster.updateTooltips = function(which) {
	if (which === "all" || which === "upgrades") {
		this.inStore = new Array(0, 0, 0, 0, 0, 0);

		Game.UpgradesById.forEach(function (upgrade, key) {
			for (var upgradeKey = 0; upgradeKey < CookieMonster.upgradeCount; upgradeKey++) {
				if (CookieMonster.checkUpgrade(upgradeKey, key, false)) {
					CookieMonster.manageTooltips(upgradeKey, key, false, false);
					break;
				}
			}
		});
	}

	if (which === "all" || which === "objects") {
		Game.ObjectsById.forEach(function (object) {
			CookieMonster.manageBuildingTooltip(object);
		});
	}
};

CookieMonster.manageTooltips = function(upgradeKey, t, n, r) {
	var i = 0;
	var s = 0;
	switch (upgradeKey) {
		case 0:
			i = this.bam('The mouse and cursors', 0.1, 0);
			break;
		case 1:
			i = this.bte(0);
			break;
		case 2:
			i = this.mcg(t);
			break;
		case 3:
			i = this.bam('Grandmas', 0.3, 1);
			break;
		case 4:
			i = this.bte(1);
			if (this.lgt(t)) {
				s++;
			}
			break;
		case 5:
			i = this.bam('Farms', 0.5, 2);
			break;
		case 6:
			i = this.bte(2);
			break;
		case 7:
			i = this.bam('Factories', 4, 3);
			break;
		case 8:
			i = this.bte(3);
			break;
		case 9:
			i = this.bam('Mines', 10, 4);
			break;
		case 10:
			i = this.bte(4);
			break;
		case 11:
			i = this.bam('Shipments', 30, 5);
			break;
		case 12:
			i = this.bte(5);
			break;
		case 13:
			i = this.bam('Alchemy labs', 100, 6);
			break;
		case 14:
			i = this.bte(6);
			break;
		case 15:
			i = this.bam('Portals', 1666, 7);
			break;
		case 16:
			i = this.bte(7);
			break;
		case 17:
			i = this.bam('Time machines', 9876, 8);
			break;
		case 18:
			i = this.bte(8);
			break;
		case 21:
			i = this.gpg();
			break;
		case 22:
			i = this.gpp();
			break;
		case 23:
			s += this.hasAchievement("Elder nap");
			if (Game.pledges === 4) {
				s += this.hasAchievement("Elder slumber");
			}
			break;
		case 24:
			s += this.hasAchievement("Elder calm");
			break;
		case 28:
			i = this.fte(1);
			break;
		case 29:
			i = this.bte(9);
			break;
		case 30:
			i = this.bam('Antimatter condensers', 99999, 9);
			break;
		case 32:
			i = this.dhc(s, t, i);
			if (this.isHeavenlyKey(t)) {
				s += this.hasAchievement("Wholesome");
			}
			break;
	}
	if (Game.UpgradesOwned === 19) {
		s += this.hasAchievement("Enhancer");
	}
	if (Game.UpgradesOwned === 49) {
		s += this.hasAchievement("Augmenter");
	}
	if (Game.UpgradesOwned === 99) {
		s += this.hasAchievement("Upgrader");
	}
	i += this.getAchievementWorth(s, t, i, 0);
	if (r) {
		return i;
	}

	return this.tooltips[t] + this.colorize(i, t, n);
};

CookieMonster.manageBuildingTooltip = function(e) {
	var t = e.id;
	var n = new Array(this.lucky("reg", true), this.lucky("frenzy", true));
	var r = new Array("none", "none");
	var o = new Array(0, 0);
	var i;

	if (Game.cookies - e.price < n[0]) {
		r[0] = "block";
		o[0] = n[0] - (Game.cookies - e.price);
	}
	if (Game.cookies - e.price < n[1]) {
		r[1] = "block";
		o[1] = n[1] - (Game.cookies - e.price);
	}

	if (e.desc === this.buildingTooltips[e.id]) {
		e.desc += '<div id="cm_ob_div_' + t + '" style="position:relative; height:96px; background:#222222; border:1px solid #000000; margin:6px -6px -6px -6px; display:none;"></div>';
		e.desc += '<div id="cm_ob_lucky_div_' + t + '" style="position:absolute; top:-25px; left:-12px; height:32px;">' + '<div id="cm_ob_lucky_div_warning" style="background:url(http://frozenelm.com/cookiemonster/images/warning.png); position:relative; float:left; height:32px; width:32px; display:none;"></div>' + '<div id="cm_ob_lucky_div_caution" style="background:url(http://frozenelm.com/cookiemonster/images/caution.png); position:relative; float:left; height:32px; width:32px; display:none;"></div>' + "</div>";
		e.desc += '<div id="cm_ob_note_div_' + t + '" style="position:absolute; left:0px; margin-top:10px; color:white;">' + '<div id="cm_ob_note_div_warning" style="background:#222222; position:relative; display:none; margin-top:4px; padding:2px; border:1px solid #FF0000;"><b style="color:#FF0000;">Warning:</b> Purchase of this item will put you under the number of Cookies required for "Lucky!"</br><span id="cm_ob_warning_amount"></span>' + '<div id="cm_ob_lucky_div_warning" style="position:absolute; left:-10px; top:-10px; height:32px; width:32px;"><img src="http://frozenelm.com/cookiemonster/images/warning.png" height=16px width=16px></div></div>' + '<div id="cm_ob_note_div_caution" style="background:#222222; position:relative; display:none; margin-top:4px; padding:2px; border:1px solid #FFFF00;"><b style="color:#FFFF00;">Caution:</b> Purchase of this item will put you under the number of Cookies required for "Lucky!" (Frenzy)</br><span id="cm_ob_caution_amount"></span>' + '<div id="cm_ob_lucky_div_warning" style="position:absolute; left:-10px; top:-10px; height:32px; width:32px;"><img src="http://frozenelm.com/cookiemonster/images/caution.png" height=16px width=16px></div></div>' + "</div>";
		Game.RebuildStore();
	}

	var u = new Array(this.colors.yellow, this.colors.yellow);
	var a = new Array(this.holdCPI[t], this.holdTC[t]);
	var f = new Array(Math.max.apply(Math, this.holdCPI), Math.max.apply(Math, this.holdTC));
	var l = new Array(Math.min.apply(Math, this.holdCPI), Math.min.apply(Math, this.holdTC));
	for (i = 0; i < u.length; i++) {
		if (a[i] === l[i]) {
			u[i] = this.colors.green;
		} else if (a[i] === f[i]) {
			u[i] = this.colors.red;
		} else if (f[i] - a[i] < a[i] - l[i]) {
			u[i] = this.colors.orange;
		}
	}

	if ($("#cm_ob_div_" + t).length === 1) {
		$("#cm_ob_div_" + t).css("border", "1px solid #" + u[0]);
		$("#cm_ob_div_" + t).css("display", "");
		$("#cm_ob_div_" + t).html('<div style="position:absolute; top:4px; left:4px; color:#4bb8f0; font-weight:bold;">Bonus Income</div><div align=right style="position:absolute; top:18px; left:4px; color:white;">' + this.formatNumber(this.holdIs[t]) + '</div><div style="position:absolute; top:34px; left:4px; color:#4bb8f0; font-weight:bold;">Base Cost Per Income</div><div align=right style="position:absolute; top:48px; left:4px; color:#' + u[0] + ';">' + this.formatNumber(a[0]) + '</div><div style="position:absolute; top:64px; left:4px; color:#4bb8f0; font-weight:bold;">Time Left</div><div align=right style="position:absolute; top:78px; left:4px; color:#' + u[1] + ';">' + this.formatTime(a[1], "") + "</div>");
		$("#cm_ob_warning_amount").text("Deficit: " + this.formatNumber(o[0]));
		$("#cm_ob_caution_amount").text("Deficit: " + this.formatNumber(o[1]));

		if (this.settings[10] === 1 || this.settings[10] === 2) {
			$("#cm_ob_lucky_div_warning").css("display", r[0]);
			$("#cm_ob_lucky_div_caution").css("display", r[1]);
		} else {
			$("#cm_ob_lucky_div_warning").css("display", "none");
			$("#cm_ob_lucky_div_caution").css("display", "none");
		}

		if (this.settings[10] === 1 || this.settings[10] === 3) {
			$("#cm_ob_note_div_warning").css("display", r[0]);
			$("#cm_ob_note_div_caution").css("display", r[1]);
		} else {
			$("#cm_ob_note_div_warning").css("display", "none");
			$("#cm_ob_note_div_caution").css("display", "none");
		}
	}

	if (this.settings[6] === 1) {
		$("#product" + t).find(".price").first().css("color", "#" + u[0]);
	} else {
		$("#product" + t).find(".price").first().css("color", "");
	}
};